Refactor config merge

eterps opened the pull request:
Refactor config merge, work in progress.
See also: https://github.com/frankroeder/parrot.nvim/discussions/22#discussioncomment-9957522

eterps commented on lua/parrot/tests/config_utils_spec.lua:
```diff
@@ -38,4 +44,69 @@ describe("config_utils", function()
+      assert.are.same({
+        {
+          name = "CustomChatGPT4",
+          model = { model = "gpt-4-turbo", temperature = 1.2 },
```
@frankroeder at this point our tests become helpful in discussing what behaviour we are expecting.

Is `model = { model = "gpt-4-turbo", temperature = 1.2 }` correct? (because it is what the user has configured)

Or should it have been `model = { model = "gpt-4-turbo", temperature = 1.2, top_p = 1 }` (the default should be merged in).

Also the first option could be considered 'invalid' (`top_p` is missing).

frankroeder replied:
All that is required is the model. This can simply be defined as a string with `model = "gpt-4-turbo"`. Multiple instances within the code make similar distinctions. This is essentially an inheritance from the fork gp.nvim.

```lua
   local user_agents = {
        {
          name = "CustomChatGPT4",
          model = "gpt-4-turbo",
          system_prompt = "You are an AI assistant",
          provider = "openai",
        },
      }
```

eterps replied:
> All that is required is the model. This can simply be defined as a string with `model = "gpt-4-turbo"`.

Thanks, in that case this test describes the expected behaviour.

eterps replied:
I discovered that `M.agents` as a data structure a bit different than the default agents list.
The config_utils function `index_agents_by_name` fixes that.

@frankroeder at this point it at least works for me. I still have to investigate what needs to be cleaned up.

eterps commented:
I discovered that `M.agents` as a data structure a bit different than the default agents list.
The config_utils function `index_agents_by_name` fixes that.

@frankroeder at this point it at least works for me. I still have to investigate what needs to be cleaned up.

frankroeder commented:
I am looking forward to merge this.

eterps commented:
> I am looking forward to merge this.

Did you look into the changes in detail? I am still unfamiliar with large parts of the code in `parrot/init.lua`. Could you give me some time to check some basic things? It won't take too long.

On the other hand, it's still merged into a separate branch, nothing will break (yet).

eterps commented:
1. Removed provider/agent filters, as the new merge functions now handle most of this functionality.
2. As a result, the current validity checks have been eliminated. We should reassess our approach to these checks.
3. Implemented a new validity check to ensure that at least one provider is configured.

eterps commented:
@frankroeder, I've removed some superfluous code.

The user config validation could still be improved, but I believe the code's readability has been enhanced overall.

Additionally, I think factoring out `M._available_providers` and `M._available_provider_agents` could further simplify the code.

Please let me know if you need any more changes. If not, feel free to merge this into your branch.
